exception when using StringConverter for key and value converter

I'm working to see if this is fixable.

```
[2016-09-28 19:16:22,211] ERROR Task coinsmith-raw-hdfs-sink-test-8-0 threw an uncaught and unrecoverable exception (org.apache.kafka.connect.runtime.WorkerSinkTask)
org.apache.avro.file.DataFileWriter$AppendWriteException: org.apache.avro.AvroRuntimeException: Unknown datum type io.confluent.kafka.serializers.NonRecordContainer: io.confluent.kafka.serializers.NonRecordContainer@31d89133
        at org.apache.avro.file.DataFileWriter.append(DataFileWriter.java:296)
        at io.confluent.connect.hdfs.avro.AvroRecordWriterProvider$1.write(AvroRecordWriterProvider.java:64)
        at io.confluent.connect.hdfs.avro.AvroRecordWriterProvider$1.write(AvroRecordWriterProvider.java:59)
        at io.confluent.connect.hdfs.TopicPartitionWriter.writeRecord(TopicPartitionWriter.java:487)
        at io.confluent.connect.hdfs.TopicPartitionWriter.write(TopicPartitionWriter.java:264)
        at io.confluent.connect.hdfs.DataWriter.write(DataWriter.java:234)
        at io.confluent.connect.hdfs.HdfsSinkTask.put(HdfsSinkTask.java:91)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:381)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:227)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:170)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:142)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:140)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:175)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.avro.AvroRuntimeException: Unknown datum type io.confluent.kafka.serializers.NonRecordContainer: io.confluent.kafka.serializers.NonRecordContainer@31d89133
        at org.apache.avro.generic.GenericData.getSchemaName(GenericData.java:636)
        at org.apache.avro.generic.GenericData.resolveUnion(GenericData.java:601)
        at org.apache.avro.generic.GenericDatumWriter.resolveUnion(GenericDatumWriter.java:151)
        at org.apache.avro.generic.GenericDatumWriter.write(GenericDatumWriter.java:71)
        at org.apache.avro.generic.GenericDatumWriter.write(GenericDatumWriter.java:58)
        at org.apache.avro.file.DataFileWriter.append(DataFileWriter.java:290)
        ... 17 more
```

--- a/src/main/java/io/confluent/connect/hdfs/avro/AvroRecordWriterProvider.java
+++ b/src/main/java/io/confluent/connect/hdfs/avro/AvroRecordWriterProvider.java
@@ -16,6 +16,7 @@
 
 package io.confluent.connect.hdfs.avro;
 
+import io.confluent.kafka.serializers.NonRecordContainer;
 import org.apache.avro.file.DataFileWriter;
 import org.apache.avro.generic.GenericDatumWriter;
 import org.apache.avro.io.DatumWriter;
@@ -61,7 +62,12 @@ public class AvroRecordWriterProvider implements RecordWriterProvider {
       public void write(SinkRecord record) throws IOException {
         log.trace("Sink record: {}", record.toString());
         Object value = avroData.fromConnectData(schema, record.value());
-        writer.append(value);
+        // AvroData wraps primitive types so their schema can be included. We need to unwrap NonRecordContainers to just
+        // their value to properly handle these types
+        if (value instanceof NonRecordContainer)
+          writer.append(((NonRecordContainer) value).getValue());
+        else
+          writer.append(value);
       }
 
       @Override

--- a/src/test/java/io/confluent/connect/hdfs/HdfsSinkTaskTest.java
+++ b/src/test/java/io/confluent/connect/hdfs/HdfsSinkTaskTest.java
@@ -185,9 +185,9 @@ public class HdfsSinkTaskTest extends TestWithMiniDFSCluster {
     Map<String, String> props = createProps();
     HdfsSinkTask task = new HdfsSinkTask();
 
-    String key = "key";
-    Schema schema = Schema.INT32_SCHEMA;
-    int record = 12;
+    final String key = "key";
+    final Schema schema = Schema.INT32_SCHEMA;
+    final int record = 12;
     Collection<SinkRecord> sinkRecords = new ArrayList<>();
     for (TopicPartition tp: assignment) {
       for (long offset = 0; offset < 7; offset++) {
@@ -217,7 +217,7 @@ public class HdfsSinkTaskTest extends TestWithMiniDFSCluster {
         long size = endOffset - startOffset + 1;
         assertEquals(records.size(), size);
         for (Object avroRecord : records) {
-          assertEquals(avroRecord, 12);
+          assertEquals(avroRecord, record);
         }
       }
     }

--- a/src/test/java/io/confluent/connect/hdfs/HdfsSinkTaskTest.java
+++ b/src/test/java/io/confluent/connect/hdfs/HdfsSinkTaskTest.java
@@ -14,6 +14,7 @@
 
 package io.confluent.connect.hdfs;
 
+import io.confluent.kafka.serializers.NonRecordContainer;
 import org.apache.hadoop.fs.Path;
 import org.apache.kafka.common.TopicPartition;
 import org.apache.kafka.connect.data.Schema;
@@ -174,6 +175,49 @@ public class HdfsSinkTaskTest extends TestWithMiniDFSCluster {
         assertEquals(records.size(), size);
         for (Object avroRecord : records) {
           assertEquals(avroRecord, avroData.fromConnectData(schema, record));
+        }
+      }
+    }
+  }
+
+  @Test
+  public void testSinkTaskPutPrimitive() throws Exception {
+    Map<String, String> props = createProps();
+    HdfsSinkTask task = new HdfsSinkTask();
+
+    String key = "key";
+    Schema schema = Schema.INT32_SCHEMA;
+    int record = 12;
+    Collection<SinkRecord> sinkRecords = new ArrayList<>();
+    for (TopicPartition tp: assignment) {
+      for (long offset = 0; offset < 7; offset++) {
+        SinkRecord sinkRecord =
+                new SinkRecord(tp.topic(), tp.partition(), Schema.STRING_SCHEMA, key, schema, record, offset);
+        sinkRecords.add(sinkRecord);
+      }
+    }
+    task.initialize(context);
+    task.start(props);
+    task.put(sinkRecords);
+    task.stop();
+
+    AvroData avroData = task.getAvroData();
+    // Last file (offset 6) doesn't satisfy size requirement and gets discarded on close
+    long[] validOffsets = {-1, 2, 5};
+
+    for (TopicPartition tp : assignment) {
+      String directory = tp.topic() + "/" + "partition=" + String.valueOf(tp.partition());
+      for (int j = 1; j < validOffsets.length; ++j) {
+        long startOffset = validOffsets[j - 1] + 1;
+        long endOffset = validOffsets[j];
+        Path path = new Path(FileUtils.committedFileName(url, topicsDir, directory, tp,
+                startOffset, endOffset, extension,
+                ZERO_PAD_FMT));
+        Collection<Object> records = schemaFileReader.readData(conf, path);
+        long size = endOffset - startOffset + 1;
+        assertEquals(records.size(), size);
+        for (Object avroRecord : records) {
+          assertEquals(avroRecord, 12);
         }
       }
     }

